lib/connections/tcp_listen.go

```diff
@@ -178,10 +178,10 @@ func (t *tcpListener) WANAddresses() []*url.URL {
 			// For every address with a specified IP, add one without an IP,
 			// just in case the specified IP is still internal (router behind DMZ).
 			if len(addr.IP) != 0 && !addr.IP.IsUnspecified() {
-				uri = *t.uri
+				zeroUri := *t.uri
 				addr.IP = nil
-				uri.Host = addr.String()
-				uris = append(uris, &uri)
+				zeroUri.Host = addr.String()
+				uris = append(uris, &zeroUri)
```

Comment:
Are you sure this makes a difference? Regardless I anyway think it's nicer the way you did it, as it's explicit that `uri` and `zeroUri` are two things. However given it dereferences `*t.uri` on assignment, it's a copy, so re-assigning `.Host` should not affect the previously added element in `uris` ("trying to repro" for demonstration: https://go.dev/play/p/CamwMcZyA89).

Reply:
The difference is that in the affected code, `uris` is an array of pointers to URLs while `l` in the repro is a list of `url`s without indirection. Therefore the repro creates a copy of `f` when calling `append(l, f)` while the original code only creates a reference to `uri` when calling `append(uris, &uri)`. A copy of the value pointed to by `t.uri` is indeed created, but that copy is written to `uri`, which is still the location which the pointer in `uris` points to. (The problem wasn't that `t.uri` was being overwritten in the first place)

If I modify the repro to do that, you can see that the array contains two equal addresses which lead to the same output when printed:

[1] https://go.dev/play/p/BPZZGjPwkVL

Reply:
Oh, what a stupid thing to miss when writing a repro. However I must admit, that was not my main misconception - I was applying a weird (and obviously wrong) assumption: For some reason I was thinking go would allocate new memory whenever a variable's value is re-assigned, thus the two appended pointers would have pointed at two different memory locations/structs.

Anyway as this isn't upnp related and an important fix, could you please PR this separately. E.g. in case we need to revert this commit, I wouldn't want to lose this fix (don't think that's going to happen, just in case).

Reply:
> Anyway as this isn't upnp related and an important fix, could you please PR this separately. E.g. in case we need to revert this commit, I wouldn't want to lose this fix (don't think that's going to happen, just in case).

I've cherry-picked the commit into #9073